### ISSUE TITLE
Error when runner catches a string error in conf.js

### ISSUE THREAD

**Issue by taras80** [angular/protractor#3224]
- Node Version: `v5.7.0`
- Protractor Version: `3.3.0`
- Output from running the test
[16:48:03] I/hosted - Using the selenium server at http://localhost:4444/wd/hub
[16:48:03] I/launcher - Running 1 instances of WebDriver
[16:48:04] E/launcher - Error: TypeError: Cannot assign to read only property 'stack' of 'here goes your error text'
    at /Users/user/projects/automation-2/node_modules/protractor/built/util.js:49:27
    at _rejected (/Users/user/projects/automation-2/node_modules/q/q.js:844:24)
    at /Users/user/projects/automation-2/node_modules/q/q.js:870:30
    at Promise.when (/Users/user/projects/automation-2/node_modules/q/q.js:1122:31)
    at Promise.promise.promiseDispatch (/Users/user/projects/automation-2/node_modules/q/q.js:788:41)
    at /Users/user/projects/automation-2/node_modules/q/q.js:604:44
    at runSingle (/Users/user/projects/automation-2/node_modules/q/q.js:137:13)
    at flush (/Users/user/projects/automation-2/node_modules/q/q.js:125:13)
    at _combinedTickCallback (node.js:370:9)
    at process._tickCallback (node.js:401:11)
[16:48:04] E/launcher - Process exited with error code 100
>>
Warning: Tests failed, protractor exited with code: 100 Use --force to continue.

Aborted due to warnings.

- Steps to reproduce the bug

1. catch errors in onPrepare of conf.js
2. throw just error string and not error object

        }).catch(function(err) {
            browser.quit();
            throw err;
        });

- As workaround you can throw error object

        }).catch(function(err) {
            browser.quit();
            throw Error(err);
        });

**Comment by mgiambalvo** [comment]
Thanks for the report. I think I see what's going on here, but just to get a little more clarity could you provide a full example config file?

**Comment by taras80** [comment]
```
/* global browser, Promise: false */

var fs = require('fs');
var HtmlReporter = require('protractor-jasmine2-screenshot-reporter');
var jasmineReporters = require('jasmine-reporters');
var xml2js = require('xml2js');

var globals = require('./globals.js');

//Protractor configuration file
exports.config = {
    //selenium configuration
    seleniumAddress: 'http://localhost:4444/wd/hub',

    //Test suite & test case composition framework
    framework: 'jasmine2',

    //Test suites definitions
    suites: {
        smokeTest: 'test-suites/smokeTest/smokeTestSpecs.js'
    },

    //Default timeout for all tests
    allScriptsTimeout: 20000,

    //Jasmine configuration
    jasmineNodeOpts: {
        includeStackTrace: true,
        isVerbose: true,
        showColors: true,
        defaultTimeoutInterval: 260000
    },

    //Define additional browser support (default is Chrome)
    capabilities: {
        browserName: 'chrome'
    },

    //Input parameters to pass to automation run
    params: {
        username: '',
        password: '',
        usernameAdmin: '',
        passwordAdmin: '',
        language: 'en',
        country: 'us',
        resultsDir: 'reports/',
        random: false,
        rerunFailed: false,
        seed: 0
    },

    //Protractor startup script (runs before execution of tests)
    onPrepare: function() {
        var rerunFailed = browser.params.rerunFailed;

        var promise = new Promise(function(resolve, reject) {
            var rerunTests = [];
            if (rerunFailed) {
                fs.readFile(rerunFailed, function(err, data) {
                    if (err) {
                        reject("Failed to read the result file. " + err);
                    } else {
                        var parser = new xml2js.Parser();
                        parser.parseString(data, function(err, parsedObj) {
                            if (err) {
                                reject("Failed to parse the result file. " + err);
                            } else if (parsedObj.testsuites && parsedObj.testsuites.testsuite) {
                                var suites = parsedObj.testsuites.testsuite;
                                suites.forEach(function(suite) {
                                    if (suite.testcase) {
                                        var cases = suite.testcase;
                                        cases.forEach(function(testcase) {
                                            if (testcase.failure) {
                                                rerunTests.push(testcase.$.name.substring(0,8));
                                            }
                                        });
                                    }
                                });
                            } else {
                                reject("Report file doesn't contain any test cases.");
                            }
                        });
                        if (rerunTests.length === 0) {
                            reject("Report file doesn't contain any failed test cases.");
                        }
                    }
                    resolve(rerunTests);
                });
            } else {
                resolve(rerunTests);
            }
        });

        return promise.then(function(rerunTests) {
            var jasmineEnv = jasmine.getEnv();

            var random = browser.params.random;
            var seed = browser.params.seed;

            var include = browser.params.include;
            var includeTests = [];

            if (include) {
                includeTests = include.toString().split(',');
            }

            //Include any failed tests that need to be rerun
            includeTests = includeTests.concat(rerunTests);

            var skip = browser.params.skip;
            var skipTests = [];

            if (skip) {
                skipTests = skip.toString().split(',');
            }

            var xmlFilePrefix = 'report';
            var htmlFilename = 'report.html';

            if (rerunFailed) {
                xmlFilePrefix = 'report-rerun';
                htmlFilename = 'report-rerun.html';
            }

            //Add test reporters to capture test results in desired formats (JUnitXML, html, Terminal)
            jasmineEnv.addReporter(new jasmineReporters.JUnitXmlReporter({
                savePath: browser.params.resultsDir,
                filePrefix: xmlFilePrefix,
                consolidate: true,
                consolidateAll: true
            }));

            jasmineEnv.addReporter(new HtmlReporter({
                dest: browser.params.resultsDir,
                filename: htmlFilename,
                showSummary: true,
                showQuickLinks: true,
                showConfiguration: false,
                captureOnlyFailedSpecs: true,
                reportOnlyFailedSpecs: false,
                ignoreSkippedSpecs: true
            }));

            jasmineEnv.addReporter(new jasmineReporters.TerminalReporter({
                verbosity: 3,
                color: true,
                showStack: false
            }));

            if (random) {
                //Run specs in random order if random is set to true or 1
                jasmineEnv.randomizeTests(random);
                //Use randomization seed if provided
                if (seed) {
                    jasmineEnv.seed(seed);
                }
            }

            //Custom spec filter to include or skip specs
            jasmineEnv.specFilter = function(spec) {
                var run = true;

                if (rerunFailed) {
                    run = false;
                }

                if (includeTests.length > 0) {
                    if (includeTests.indexOf(spec.description.substring(0,8)) !== -1) {
                        run = true;
                    } else {
                        run = false;
                    }
                } else if (skipTests.length > 0) {
                    if (skipTests.indexOf(spec.description.substring(0,8)) !== -1 ) {
                        run = false;
                    }
                }

                return run;
            };

            //Mac Mini resolution
            browser.manage().window().setSize(1280, 1440);

        }).catch(function(err) {
            browser.quit();
            throw Error(err);
        });
    }
};

```

**Comment by cnishina** [comment]
`onPrepare` method runs with the util method [runFilenameOrFn_](https://github.com/angular/protractor/blob/master/lib/util.ts#L50). 

When throwing  an error from `onPrepare`, it is being caught on the runFilenameOrFn_ method. The runFilenameOrFn_ method uses the stack trace from the error object. Since there is no stack trace given from a Promise.reject call, there are a couple workaround options:

1. Just print the message on catch and not throw the error message since it does not have a stack.
2. Instead of using the `reject('here goes your text')`, use `throw new Error('here goes your text')`. In your catch method, you can still throw the error since it has a stack trace on the created error.

This is also an easy fix and we could easily check if stack exists before accessing stack. (https://github.com/angular/protractor/blob/master/lib/util.ts#L52)

**Comment by cnishina** [comment]
There might be a problem with stack traces not being generated. Potentially "fixing" the problem is "hiding" the problem.

**Comment by cnishina** [comment]
Reopening this for discussion.

**Comment by juliemr** [comment]
If node does not attach a stack to a thrown string, we should go through and make sure that we always check for the stack property on errors before modifying.

**Comment by cnishina** [comment]
Should be addressed with 9182d58a5d3e9d6210f4f12144e56cc407510440.
